src/cursor/cur_version.c

```diff
@@ -0,0 +1,470 @@
+             * version cursor every time we traverse a version, we can simply apply the modify onto
+             * the latest value.
+             */
+            if (hs_upd_type == WT_UPDATE_MODIFY) {
+                WT_ERR(__wt_modify_apply_item(
+                  session, table_cursor->value_format, &cursor->value, hs_value.data));
+            } else {
+                WT_ASSERT(session, hs_upd_type == WT_UPDATE_STANDARD);
+                cursor->value.data = hs_value.data;
+                cursor->value.size = hs_value.size;
+            }
+            F_SET(cursor, WT_CURSTD_VALUE_INT);
+            upd_found = true;
+        } else
+            F_SET(version_cursor, WT_VERSION_CUR_HS_EXAUSTED);
+    }
+
+done:
+    if (0) {
+err:
+        WT_TRET(cursor->reset(cursor));
+    }
+    API_END_RET(session, ret);
+}
+
+/*
+ * __curversion_reset --
+ *     WT_CURSOR::reset for version cursors.
+ */
+static int
+__curversion_reset(WT_CURSOR *cursor)
+{
+    WT_CURSOR *hs_cursor, *table_cursor;
+    WT_CURSOR_VERSION *version_cursor;
+    WT_DECL_RET;
+    WT_SESSION_IMPL *session;
+
+    version_cursor = (WT_CURSOR_VERSION *)cursor;
+    hs_cursor = version_cursor->hs_cursor;
+    table_cursor = version_cursor->table_cursor;
+    CURSOR_API_CALL(cursor, session, reset, NULL);
+
+    if (table_cursor != NULL)
+        WT_TRET(table_cursor->reset(table_cursor));
+    if (hs_cursor != NULL)
+        WT_TRET(hs_cursor->reset(hs_cursor));
+    version_cursor->next_upd = NULL;
+    version_cursor->flags = 0;
+    F_CLR(cursor, WT_CURSTD_KEY_SET);
+    F_CLR(cursor, WT_CURSTD_VALUE_SET);
+
+err:
+    API_END_RET(session, ret);
+}
+
+/*
+ * __curversion_search --
+ *     WT_CURSOR->search method for version cursors.
+ */
+static int
+__curversion_search(WT_CURSOR *cursor)
+{
+    WT_COL *cip;
+    WT_CURSOR *table_cursor;
+    WT_CURSOR_BTREE *cbt;
+    WT_CURSOR_VERSION *version_cursor;
+    WT_DECL_RET;
+    WT_INSERT *ins;
+    WT_PAGE *page;
+    WT_ROW *rip;
+    WT_SESSION_IMPL *session;
+    WT_UPDATE *upd;
+    bool key_only;
+
+    version_cursor = (WT_CURSOR_VERSION *)cursor;
+    table_cursor = version_cursor->table_cursor;
+    key_only = F_ISSET(cursor, WT_CURSTD_KEY_ONLY);
+
+    /*
+     * For now, we assume that we are using simple cursors only.
+     */
+    cbt = (WT_CURSOR_BTREE *)table_cursor;
+    CURSOR_API_CALL(cursor, session, search, CUR2BT(cbt));
+    WT_ERR(__cursor_checkkey(table_cursor));
+
+    /* Do a search and position on they key if it is found */
+    F_SET(cursor, WT_CURSTD_KEY_ONLY);
+    WT_ERR(__wt_btcur_search(cbt));
+    WT_ASSERT(session, F_ISSET(cbt, WT_CURSTD_KEY_SET));
+
+    /*
+     * If we position on a key, set next update of the version cursor to be the first update on the
+     * key if any.
+     */
+    page = cbt->ref->page;
+    switch (page->type) {
+    case WT_PAGE_ROW_LEAF:
+        rip = &page->pg_row[cbt->slot];
```

Comment:
if (cbt->ins != NULL) {
} else {
    rip = &page->pg_row[cbt->slot];
    upd = WT_ROW_UPDATE(page, rip)) 
    if (upd != NULL) {
    }
}

Reply:
I will also simplify the logic for both the row store and column store sections, I don't think I need to be checking NULLs since this is already done in the next() logic.